File: csharp/Memento.cs
/// @file
/// @brief
/// The @ref DesignPatternExamples_csharp.Memento_TextObject "Memento_TextObject"
/// class used in the @ref memento_pattern "Memento pattern".

namespace DesignPatternExamples_csharp
{
    /// <summary>
    /// Represents a single memento, a single snapshot of the state of the
    /// Memento_TextObject class as represented by the private class
    /// Memento_TextObject.Memento.
    /// </summary>
    public interface IMemento
    {
        /// <summary>
        /// The name of the memento (snapshot).  Useful for displaying a list
        /// of mementos in an undo list.  In this case, the name of each
        /// memento is the operation that triggered the creation of the
        /// memento.
        /// </summary>
        string Name { get; }
    }


    //########################################################################
    //########################################################################


    /// <summary>
    /// Container for a string.  Need to use a class that allows the text to
    /// be changed while the container (this class) remains constant.  This
    /// way, operations can be applied to the text and the container's contents
    /// change but not the container (strings are immutable in C#; this dances
    /// around that problem).
    /// </summary>
    public class Memento_TextObject
    {
        //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
        // Inner class Memento

        /// <summary>
        /// Represents a single memento (snapshot) of the text state before
        /// an operation is applied.  The operation becomes the name of the
        /// memento for display purposes.
        /// </summary>
        internal class Memento : IMemento
        {
            /// <summary>
            /// The name of this memento (really just the name of the operation
            /// that triggered the need for this memento).
            /// </summary>
            string _name;

            /// <summary>
            /// The snapshot of the text data as stored in the Memento_TextObject
            /// class instance.
            /// </summary>
            string _text;


            /// <summary>
            /// Constructor.
            /// </summary>
            /// <param name="name">The name of the memento to create.</param>
            /// <param name="text">The data to be saved in the memento</param>
            internal Memento(string name, string text)
            {
                _name = name;
                _text = text;
            }


            /// <summary>
            /// The saved text in this memento.  This is accessible only by the
            /// Memento_TextObject class since it is the only entity that knows
            /// what to do with the text during an undo.
            /// </summary>
            internal string Text
            {
                get
                {
                    return _text;
                }
            }

            /// <summary>
            /// The name of this memento.  This is seen as the operation that
            /// triggered the need for the memento.
            /// </summary>
            string IMemento.Name
            {
                get
                {
                    return _name;
                }
            }
        }

        // End of Inner class Memento
        //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
        //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


        //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
        // Private data of the Memento_TextObject class.

        /// <summary>
        /// The text that can change in this Memento_TextObject class.
        /// </summary>
        string _text;

        // End private data of the Memento_TextObject class.
        //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
        //++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


        /// <summary>
        /// Constructs a text object with an initial string.
        /// </summary>
        /// <param name="text">The initial string for the text object.</param>
        public Memento_TextObject(string text)
        {
            _text = text;
        }

        /// <summary>
        /// Gets or sets the text in this TextObject.
        /// </summary>
        public string Text
        {
            get
            {
                return _text;
            }
            set
            {
                _text = value;
            }
        }

        /// <summary>
        /// Returns an IMemento object containing a snapshot of the text
        /// stored in this class instance.
        /// </summary>
        public IMemento GetMemento(string operationName)
        {
            return new Memento(operationName, _text);
        }


        /// <summary>
        /// Sets the text in this class instance to the snapshot stored in the
        /// given IMemento object (which is assumed to be from this class).
        /// </summary>
        /// <param name="memento">The IMemento object to restore to.</param>
        public void RestoreMemento(IMemento memento)
        {
            _text = ((Memento)memento).Text;
        }


        /// <summary>
        /// Converts the Memento_TextObject to a string (makes it easier to
        /// use the class in string formatting).
        /// </summary>
        /// <returns>Returns the text stored in the text object.</returns>
        public override string ToString()
        {
            return _text;
        }
    }
}

File: python/DesignPatternExamples_python/memento/memento.py
## @file
# @brief
# Implementation of the
# @ref DesignPatternExamples_python.memento.memento.Memento_TextObject "Memento_TextObject"
# and (private)
# @ref DesignPatternExamples_python.memento.memento.Memento_TextObject.Memento "Memento"
# class, along with
# @ref DesignPatternExamples_python.memento.memento.IMemento "IMemento"
# interface implemented on the Memento class, as used in the @ref memento_pattern.

from abc import ABC, abstractmethod


## Represents a single memento, a single snapshot of the state of the
#  @ref DesignPatternExamples_python.memento.memento.Memento_TextObject "Memento_TextObject"
#  class as represented by the private class,
#  @ref DesignPatternExamples_python.memento.memento.Memento_TextObject.Memento "Memento".
class IMemento(ABC):
    ## Property getter for the name of the memento (snapshot): `value = o.Name`.
    #  Useful for displaying a list of mementos in an undo list.  In this case,
    #  the name of each memento is the operation that triggered the creation of
    #  the memento.
    @abstractmethod
    @property
    def Name(self) -> str:
        pass


#########################################################################
#########################################################################


##  Container for a string.  Need to use a class that allows the text to
#  be changed while the container (this class) remains constant.  This
#  way, operations can be applied to the text and the container's contents
#  change but not the container (strings are immutable in C#; this dances
#  around that problem).
class Memento_TextObject:
    #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
    # Inner class Memento

    ## Represents a single memento (snapshot) of the text state before
    #  an operation is applied.  The operation becomes the name of the
    #  memento for display purposes.
    class Memento(IMemento):
        ## Constructor.
        #
        #  @param name
        #         The name of the memento to create.
        #  @param text
        #         The data to be saved in the memento
        def __init__(self, name : str, text : str) -> None:
            self._name = name
            self._text = text

        ## @var _name
        #       The name of this memento (really just the name of the operation
        #       that triggered the need for this memento).
        ## @var _text
        #       The snapshot of the text data as stored in the
        #       @ref DesignPatternExamples_python.memento.memento.Memento_TextObject "Memento_TextObject"
        #       class instance.


        ## Property getter for the saved text in this memento: `value = o.Text`.
        #  This is accessible only by the
        #  @ref DesignPatternExamples_python.memento.memento.Memento_TextObject "Memento_TextObject"
        #  class since it is the only entity that knows what to do with the
        #  text during an undo.
        @property
        def Text(self) -> str:
            return self._text

        ## Property getter for the name of this memento: `value = o.Name`.
        #  This is seen as the operation that triggered the need for the
        #  memento.
        @property
        def Name(self) -> str:
            return self._name

    # End of Inner class Memento
    #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
    #++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


    ## @name Properties
    #  @{

    ## Property getter for the text in this TextObject: `value = o.Text`.
    @property
    def Text(self) -> str:
        return self._text


    ## Property setter for the text in this TextObject: `o.Text = value`
    @Text.setter
    def Text(self, value : str) -> None:
        self._text = value;

    ## @}

    ## Constructs a text object with an initial string.
    #
    #  @param text
    #         The initial string for the text object.
    def __init__(self, text : str) -> None:
        self._text = text;

    ## @var _text
    #       The text that can change in this
    #       @ref DesignPatternExamples_python.memento.memento.Memento_TextObject "Memento_TextObject"
    #       class.

    ## Returns an
    #  @ref DesignPatternExamples_python.memento.memento.IMemento "IMemento"
    #  object containing a snapshot of the text stored in this class instance.
    #
    #  @param operationName
    #         Name of the memento (the operation associated with this memento).
    def GetMemento(self, operationName : str) -> IMemento:
        return Memento_TextObject.Memento(operationName, self._text)


    ## Sets the text in this class instance to the snapshot stored in the
    #  given IMemento object (which is assumed to be from this class).
    #
    #  @param memento
    #         The IMemento object to restore to.
    def RestoreMemento(self, memento : IMemento) -> None:
        if memento:
            self._text = memento.Text


    ## Converts the Memento_TextObject to a string (makes it easier to
    #  use the class in string formatting).
    #
    #  @returns
    #     Returns the text stored in the text object.
    def ToString(self) -> str:
        return self._text;

